File: backend-go/internal/service/result_mapper.go
package service

import (
	"fmt"

	"github.com/pdfdiff/backend-go/internal/domain"
	"github.com/pdfdiff/backend-go/internal/dto"
)

func BuildCompareResult(
	jobID string,
	templateLines, contractLines []domain.LineUnit,
	rawChanges []domain.RawChange,
) dto.CompareResult {
	var changes []dto.ChangeItem
	summary := dto.CompareSummary{}
	changeIndex := 0

	for _, raw := range rawChanges {
		if raw.Type == "equal" {
			summary.EqualLines++
			continue
		}

		changeIndex++
		item := toChangeItem(fmt.Sprintf("c%04d", changeIndex), raw)
		changes = append(changes, item)
		summary = updateSummary(summary, item)
	}

	templateInfo := make([]dto.LineInfo, len(templateLines))
	for i, line := range templateLines {
		templateInfo[i] = toLineInfo(line)
	}
	contractInfo := make([]dto.LineInfo, len(contractLines))
	for i, line := range contractLines {
		contractInfo[i] = toLineInfo(line)
	}

	return dto.CompareResult{
		JobID:         jobID,
		Status:        "done",
		Summary:       summary,
		Changes:       changes,
		TemplateLines: templateInfo,
		ContractLines: contractInfo,
	}
}

func toChangeItem(changeID string, raw domain.RawChange) dto.ChangeItem {
	return dto.ChangeItem{
		ID:       changeID,
		Type:     raw.Type,
		Level:    "line",
		Template: sideFromLines(raw.TemplateLines),
		Contract: sideFromLines(raw.ContractLines),
	}
}

func sideFromLines(lines []domain.LineUnit) *dto.SideInfo {
	if len(lines) == 0 {
		return nil
	}
	line := lines[0]
	bbox := line.BBox
	return &dto.SideInfo{
		Page:   line.Page,
		Text:   line.Text,
		BBoxes: [][]float64{{bbox[0], bbox[1], bbox[2], bbox[3]}},
	}
}

func toLineInfo(line domain.LineUnit) dto.LineInfo {
	bbox := line.BBox
	return dto.LineInfo{
		ID:     line.ID,
		Page:   line.Page,
		Text:   line.Text,
		BBoxes: [][]float64{{bbox[0], bbox[1], bbox[2], bbox[3]}},
	}
}

func updateSummary(summary dto.CompareSummary, item dto.ChangeItem) dto.CompareSummary {
	switch item.Type {
	case "delete":
		summary.DeletedLines++
	case "insert":
		summary.InsertedLines++
	case "replace":
		summary.ModifiedLines++
	}
	return summary
}

File: backend-node/src/lib/compare/resultMapper.ts
import type { ChangeItem, CompareResult, LineInfo, PdfPageSize } from '../../types/compare.js'
import type { LineUnit, RawChange } from './types.js'

function sideFromLines(
  lines: LineUnit[],
  bboxesOverride: number[][] | null,
): ChangeItem['template'] {
  if (!lines.length || !bboxesOverride?.length) return null
  const line = lines[0]
  return { page: line.page, text: line.text, bboxes: bboxesOverride }
}

function toChangeItem(changeId: string, raw: RawChange): ChangeItem {
  return {
    id: changeId,
    type: raw.type as ChangeItem['type'],
    level: 'char',
    template: sideFromLines(raw.templateLines, raw.templateBboxes),
    contract: sideFromLines(raw.contractLines, raw.contractBboxes),
  }
}

function toLineInfo(line: LineUnit): LineInfo {
  return {
    id: line.id,
    page: line.page,
    text: line.text,
    bboxes: [line.bbox],
  }
}

export function buildCompareResult(
  jobId: string,
  templateLines: LineUnit[],
  contractLines: LineUnit[],
  rawChanges: RawChange[],
  templatePageSizes?: PdfPageSize[],
  contractPageSizes?: PdfPageSize[],
): CompareResult {
  const changes: ChangeItem[] = []
  let deletedLines = 0
  let insertedLines = 0
  let modifiedLines = 0
  let equalLines = 0
  let changeIndex = 0

  for (const raw of rawChanges) {
    if (raw.type === 'equal') {
      equalLines++
      continue
    }

    changeIndex++
    const item = toChangeItem(`c${String(changeIndex).padStart(4, '0')}`, raw)
    changes.push(item)

    if (item.type === 'delete') deletedLines++
    else if (item.type === 'insert') insertedLines++
    else if (item.type === 'replace') modifiedLines++
  }

  return {
    job_id: jobId,
    status: 'done',
    summary: {
      deleted_lines: deletedLines,
      inserted_lines: insertedLines,
      modified_lines: modifiedLines,
      equal_lines: equalLines,
    },
    changes,
    template_lines: templateLines.map(toLineInfo),
    contract_lines: contractLines.map(toLineInfo),
    template_page_sizes: templatePageSizes,
    contract_page_sizes: contractPageSizes,
  }
}
